proposals/pattern-const-expressions.md

```diff
@@ -0,0 +1,97 @@
+# Constant `is` pattern expressions
+
+* [x] Proposed
+* [ ] Prototype: None
+* [ ] Implementation: None
+* [ ] Specification: Started, below
+
+## Summary
+[summary]: #summary
+
+Consider `is` pattern expressions as constant if the LHS is constant.
+
+## Motivation
+[motivation]: #motivation
+
+When the pattern expressions are converted into simpler boolean expressions using equality and comparison operators, they are considered constant expressions if all parts of the expression are considered constant and can be evaluated during compilation. Also, the expressions returned from the ternary operator can also be considered constant if all sides are constant (condition, consequence and alternative).
+
+## Detailed design
+[design]: #detailed-design
+
+### Spec
+The [section](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1223-constant-expressions) in the spec needs to be updated accordingly, in the following segments:
+
+> Only the following constructs are permitted in constant expressions:
+> - Literals (including the null literal).
+> - References to const members of class and struct types.
+> - References to members of enumeration types.
+> - References to local constants.
+> - Parenthesized subexpressions, which are themselves constant expressions.
+> - Cast expressions.
+> - checked and unchecked expressions.
+> - nameof expressions.
+> - The predefined +, –, !, and ~ unary operators.
+> - The predefined +, –, *, /, %, <<, >>, &, |, ^, &&, ||, ==, !=, <, >, <=, and >= binary operators.
+> - The ?: conditional operator.
+> - **Pattern expressions.**
```

Comment:
This is nowhere close to enough info. What types of patterns are allowed? What conversions, if any, are allowed? This needs to be significantly expanded.

Reply:
As I see it, we should include:
- numeric literal patterns
- null literal patterns
- default literal patterns
- relative numeric literal patterns
- char literal patterns
- string literal patterns
- type check patterns without member (field/property) evaluators

Excluded are:
- patterns with member evaluators (including type check or not)
- `var` patterns
- list patterns

On all of the above included patterns, the LHS must always be a constant.

As for the constant patterns, we would not mind the ability to corrupt the memory via unsafe methods, invalidating the constant expression's result, resulting in an inconsistency (i.e. changing the value of a const field that causes an expression to return true instead of false).

Reply:
I generally agree with the list, but not the type pattern check. We don't have any analogous constant check to that today, and it doesn't seem useful to me.